coordinate-systems.md

```diff
@@ -0,0 +1,246 @@
+
+Currently, XRFrameOfReference refers to an enumerated set of standard coordinate systems relative to which all others must be able to be expressed. These include:
+
+
+
+*   _eye-level_: the origin and rotation of this frame of reference are the orientation of the device at the start of session.
+*   _stage_: the origin and rotation of this frame of reference are in the center of a known safe playing area like a VR room.
+*   _head-model_: this frame of reference is like eye-level except the translation component of the device pose is basically identity with a neck model. This is for experiences which expect the user to not move and only look around.
+
+These are basically different ways of expressing the transform of the device in 3D space. For example, virtual objects would have the same transform in _eye-level_, _stage_, and _head-model_ FORs, it's only the device that behaves differently (more on this later). _Stage_ basically allows for a virtual translation to be applied to the native device transform and _head-model_ always locks the device translation to (0, 0, 0).
+
+Finally, XRCoordinateSystem was not designed to be dynamic. If we want to use XRCoordinateSystems to track moving objects or reference frames, we need to design a system/API to handle that dynamicism. If we want to be able to express things like hit-test input rays relative to a device or input-source, we need to be able to represent that relationship. We also need to know what coordinate system the results are in.
+
+
+# Analysis
+
+
+## XRFrameOfReference
+
+
+### Basics
+
+The current API for using XRFrameOfReference (FOR) is as follows:
+
+
+```
+const frameOfReference = await session.requestFrameOfReference('stage');
+```
+
+
+Now that we have the frame of reference we want, we convert any XRCoordinateSystem into 'stage-space' by transforming by that FOR. Most simply, if we want the current transform of the device, we say:
+
+
+```
+let devicePose = xrFrame.getDevicePose(frameOfReference);
+```
+
+
+This is the pose of the device in the given FOR. If you want to think in terms of 'absolute space', the origin of all 3 styles of FORs is the start-of-session pose of the device - that is the device pose when the AR system was first activated. In fact, the documentation around the method XRFrameOfReference::TransformBasePose(TransformationMatrix base_pose) explains this well: "Transforms a given pose from a "base" coordinate system used by the XR service to the frame of reference's coordinate system." What this means is pretty simple - the conceit is that most transforms will naturally be expressed in some native system that has its own reference frame - in most cases it is relative to the start-of-service of the underlying AR/VR runtime.
```

Comment:
(I think this is what I was saying above, thanks @thetuvix )

Reply:
This is a great point I had not considered. I am interested in learning more about how this works in practice with respect to the FOR types. Also, does this have any meaning in VR or is it only for AR?

Reply:
@thetuvix what happens if, after walking through the dark hallway, you turn around and look back at where you were? Would you expect that none of the objects in the first reference frame are now visible - i.e. we create completely disjoint spaces somehow? If they would be visible, it seems like they would be in wildly incorrect locations and it seems like we don't have a strong abstraction to handle virtual objects in disconnected coordinate systems.

Reply:
A `"stage"` frame tends to be surfaced more for VR devices today, although nothing stops an AR device from finding a similarly open place on the floor in a room and exposing it as a `"stage"`.

In general, I find the terms "VR" and "AR" to be quite overloaded.  For example, whether the platform can find a "stage" relates to its tracking of the world around you, which is mostly orthogonal to the device's opacity - that's why I like the `environmentBlendMode` term, which makes it quite clear what slice of "AR-ness" is being communicated.

Reply:
Regarding looking back down a dark hallway, if the app does the naïve thing and just leaves an object on an anchor that lost tracking loose in the engine's world coordinates (having last updated those coordinates when the anchor did have tracking), the object will indeed be in the wrong place, having "dragged along" approximately the distance the user traveled during tracking loss.

It is the responsibility of an app (or ideally, a world-space-aware engine) to hide the tree of content under an anchor during frames in which that anchor cannot be located, rather than just leaving it in its last-known location.